### PR TITLE
feat: Add deps & test in cargo

### LINKED ISSUE #5: Translate cargo
https://github.com/nebula-labs/rust-by-example-vn/tree/main/src/cargo

1. build_scripts.md (Duong)
2. conventions.md (Duong)
3. deps.md (Dung)
4. test.md (Dung)
5. [raw_identifiers.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/compatibility/raw_identifiers.md) (Vu)

### PR DESCRIPTION
**Pull request by frenkie-ng**
Closes: #5 

1.     Translate src/cargo/deps.md
2.     Translate src/cargo/test.md

### CODE REVIEW

File: src/cargo/deps.md [hunk 1]
```diff
@@ -1 +1,66 @@
 # Dependencies
+
+Hầu hết các chương trình đều phụ thuộc vào các thư viện. Nếu bạn đã từng tự tay quản lí các dependency, bạn biết nó vất vả như thế nào. Thật may mắn, hệ sinh thái ngôn ngữ Rust được chuẩn hóa với `cargo`! `cargo` có thể quản lí các dependency cho một dự án.
```
Comment (HongThaiPham) [inline src/cargo/deps.md]:
```suggestion
Hầu hết các chương trình đều phụ thuộc vào một vài thư viện nào đó. Nếu bạn đã từng tự tay quản lí các thư viện phụ thuộc này, bạn sẽ biết việc này có thể rất phiền toái. May mắn thay, hệ sinh thái Rust đi kèm với công cụ cargo! cargo có thể quản lý các phụ thuộc cho một dự án.
```

File: src/cargo/deps.md [hunk 2]
```diff
@@ -1 +1,66 @@
...
+
+Để tạo mới một project Rust,
+
+```rust,mdbook-runnable,editable
+# tạo nhị phân
```
Comment (HongThaiPham) [inline src/cargo/deps.md]:
```suggestion
# một tệp nhị phân
```

File: src/cargo/deps.md [hunk 3]
```diff
@@ -1 +1,66 @@
...
+cargo new foo
+
+# tạo thư viện
+cargo new --lib bar
+```
+
+Với phần còn lại của chương này, giả sử chúng ta đang tạo 1 hệ nhị phân, hơn là một thư viện, song mọi khái niệm là giống nhau.
```
Comment (HongThaiPham) [inline src/cargo/deps.md]:
```suggestion
Trong phần tiếp theo của chương này, giả sử rằng chúng ta đang tạo 1 tập nhị phân, chứ không phải một thư viện, song các khái niệm là giống nhau.
```

File: src/cargo/deps.md [hunk 4]
```diff
@@ -1 +1,66 @@
...
+├── bar
+│   ├── Cargo.toml
+│   └── src
+│       └── lib.rs
+└── foo
+    ├── Cargo.toml
+    └── src
+        └── main.rs
+```
+`main.rs` là tệp tin gốc của project `foo` vừa được tạo.`Cargo.toml` là tệp tin chứa cấu hình cho `cargo` của project. Nếu bạn nhìn vào bên trong, bạn sẽ thấy một thứ như thế này:
```
Comment (HongThaiPham) [inline src/cargo/deps.md]:
```suggestion
`main.rs` là tệp tin gốc của dự án `foo` vừa được tạo - không có gì mới ở đây. `Cargo.toml` là tệp tin chứa cấu hình của `cargo` đối với dự án này. Nếu bạn xem bên trong tệp tin, bạn sẽ thấy giống như thế này: 
```

File: src/cargo/deps.md [hunk 5]
```diff
@@ -1 +1,66 @@
...
+`main.rs` là tệp tin gốc của project `foo` vừa được tạo.`Cargo.toml` là tệp tin chứa cấu hình cho `cargo` của project. Nếu bạn nhìn vào bên trong, bạn sẽ thấy một thứ như thế này: 
+```rust
+[package]
+name = "foo"
+version = "0.1.0"
+authors = ["mark"]
+
+[dependencies]
+```
+Trường `name` bên dưới `[package]` xác định tên của project.Tên project cũng được sử dụng nếu bạn xuất bản crate lên `crates.io`. Nó cũng là tên của tệp nhị phân đầu ra khi được biên dịch.
```
Comment (HongThaiPham) [inline src/cargo/deps.md]:
```suggestion
Trường `name` bên dưới `[package]` xác định tên của dự án. Tên dự án cũng được sử dụng nếu bạn xuất bản crate lên `crates.io`. Nó cũng là tên của tệp nhị phân đầu ra khi được biên dịch.
```

File: src/cargo/deps.md [hunk 6]
```diff
@@ -1 +1,66 @@
...
+
+Trường `version` xác định phiên bản của crate theo quy chuẩn [Semantic Versioning](https://semver.org/).
+
+Trường `authors` là một danh sách các tác giả được sử dụng khi xuất bản crate.
+
+Phần `[dependencies]` để bạn tùy ý thêm các dependency vào project của mình.
```
Comment (HongThaiPham) [inline src/cargo/deps.md]:
```suggestion
Phần `[dependencies]` để bạn tùy ý thêm các thư viện phụ thuộc vào dự án của mình.
```

File: src/cargo/deps.md [hunk 7]
```diff
@@ -1 +1,66 @@
...
+
+Ví dụ, giả sử rằng chúng ta muốn chương trình của mình có một CLI tuyệt vời. Bạn có thể tìm kiếm rất nhiều package trên [crates.io](https://crates.io/) (Đăng ký chính thức của Rust). Một lựa chọn phổ biến là [clap](https://crates.io/crates/clap). Khi viết bài này, phiên bản `clap`được công bố gần đây nhất là `2.27.1`. Để thêm một depencency tới chương trình của mình, chúng ta đơn giản chỉ cần thêm vào tệp `Cargo.toml` ở phía dưới phần `dependencies`:`clap = "2.27.1"`.  Và thế là xong. Bạn có thể bắt đầu sử dụng `clap` trong chương trình của mình.
```
Comment (HongThaiPham) [inline src/cargo/deps.md]:
```suggestion
Ví dụ, giả sử rằng chúng ta muốn chương trình của mình có một CLI tuyệt vời. Bạn có thể tìm kiếm thấy rất nhiều package trên [crates.io](https://crates.io/) (Địa chỉ chính thức của hệ thống đăng ký gói của Rust). Một lựa chọn phổ biến là [clap](https://crates.io/crates/clap). Tại thời điểm viết bài này, phiên bản mới nhất của `clap` được công bố là `2.27.1`. Để thêm một phục thuộc cho chương trình của mình, chúng ta đơn giản chỉ cần thêm vào tệp `Cargo.toml` ở phía dưới phần `dependencies`:`clap = "2.27.1"`.  Và thế là xong. Bạn có thể bắt đầu sử dụng `clap` trong chương trình của mình.
```

File: src/cargo/deps.md [hunk 8]
```diff
@@ -1 +1,66 @@
...
+```rust
+[package]
+name = "foo"
+version = "0.1.0"
+authors = ["mark"]
+
+[dependencies]
+clap = "2.27.1" # from crates.io
+rand = { git = "https://github.com/rust-lang-nursery/rand" } # từ repo online.
+bar = { path = "../bar" } # từ một tệp dẫn trong hệ thống tệp cục bộ.
```
Comment (HongThaiPham) [inline src/cargo/deps.md]:
```suggestion
bar = { path = "../bar" } # đường dẫn trong hệ thống tệp cục bộ.
```

File: src/cargo/deps.md [hunk 9]
```diff
@@ -1 +1,66 @@
...
+```
+`cargo` không chỉ là một trình quản lí các dependency. Tất cả các tùy chọn cấu hình đều có sẵn và được liệt kê trong [đặc tả định dạng](https://doc.rust-lang.org/cargo/reference/manifest.html) của `Cargo.toml`
```
Comment (HongThaiPham) [inline src/cargo/deps.md]:
```suggestion
`cargo` không chỉ là một trình quản lí các phụ thuộc. Tất cả các tùy chọn cấu hình sẵn có đều được liệt kê trong [đặc tả định dạng](https://doc.rust-lang.org/cargo/reference/manifest.html) của `Cargo.toml`
```

File: src/cargo/deps.md [hunk 10]
```diff
@@ -1 +1,66 @@
...
+
+Để  build project của mình chúng ta có thể  thực thi bằng câu lệnh `cargo build` ở bất kì đâu trong thư mục project ( ngay cả trong các thư mục con). Chúng ta có thể chạy `cargo run` để build và run. Lưu ý rằng những câu lệnh này sẽ giải quyết tất cả các dependency, tải các crate nếu cần, và build tất cả mọi thứ, bảo gồm crate của mình. ( Lưu ý rằng nó chỉ build lại những gì mà nó chưa được build, tương tự như `make`).
```
Comment (HongThaiPham) [inline src/cargo/deps.md]:
```suggestion
Để  build dự án, chúng ta có thể  thực thi lệnh `cargo build` ở bất kì đâu trong thư mục dự án (ngay cả trong các thư mục con). Chúng ta có thể sử dụng `cargo run` để build và chạy chương trình. Lưu ý rằng những câu lệnh này sẽ giải quyết tất cả các phụ thuộc, tải xuống các crate nếu cần, và build tất cả mọi thứ, bao gồm crate của bạn. ( Lưu ý rằng nó chỉ build lại những gì mà nó chưa được build trước đó, tương tự như `make`).
```

File: src/cargo/deps.md [hunk 11]
```diff
@@ -1 +1,66 @@
...
+
+Thì đấy! Đó là tất cả về dependency.
```
Comment (HongThaiPham) [inline src/cargo/deps.md]:
```suggestion
Trên đây là tất cả về quản lý các phụ thuộc.
```

File: src/cargo/test.md [hunk 12]
```diff
@@ -1 +1,134 @@
 # Tests
+Như chúng ta đã biết kiểm thử  không thể thiếu đối với bất kì giai đoạn nào trong quá trình phát triển phần mềm. Rust ưu tiên cho việc unit test( kiểm thử đơn vị) và integration test( kiểm thử tích hợp) ( [xem chương này](https://doc.rust-lang.org/book/ch11-00-testing.html) tại TRPL).
+
+Từ các chương kiểm thử được liên kết phía trên, chúng ta sẽ biết cách để viết unit test và integration test.
```
Comment (HongThaiPham) [inline src/cargo/test.md]:
```suggestion
Từ các chương kiểm thử trong liên kết phía trên, chúng ta sẽ biết cách để viết unit test và integration test.
```

File: src/cargo/test.md [hunk 13]
```diff
@@ -1 +1,134 @@
...
+Về mặt tổ chức, chúng tôi đặt các unit test bên trong các mô-đun, chúng sẽ test và integration test trong chính thư mục `tests/`:
```
Comment (HongThaiPham) [inline src/cargo/test.md]:
```suggestion
Về mặt tổ chức, chúng ta đặt các unit test bên trong các module, chúng sẽ chạy kiểm tra và tích hợp kiểm thử trong chính thư mục `tests/`:
```

File: src/cargo/test.md [hunk 14]
```diff
@@ -1 +1,134 @@
...
+├── Cargo.toml
+├── src
+│   └── main.rs
+│   └── lib.rs
+└── tests
+    ├── my_test.rs
+    └── my_other_test.rs
+
+```
+Mỗi tệp tin trong `tests` là một [integration test](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests) riêng biệt, tức là kiểm thử đó nhằm kiểm tra thư viện của bạn như thể nó được gọi từ một dependency crate.
```
Comment (HongThaiPham) [inline src/cargo/test.md]:
```suggestion
Mỗi tệp tin trong `tests` là một [integration test](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests) riêng biệt, tức là kiểm thử đó nhằm kiểm tra thư viện của bạn như thể nó được gọi từ một crate phụ thuộc. 
```

File: src/cargo/test.md [hunk 15]
```diff
@@ -1 +1,134 @@
...
+running 3 tests
+test test_bar ... ok
+test test_baz ... ok
+test test_foo_bar ... ok
+test test_foo ... ok
+
+test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
+
+```
+Bạn cũng có thể chạy kiểm thử với tên test cụ thể(`test_foo`):
```
Comment (HongThaiPham) [inline src/cargo/test.md]:
```suggestion
Bạn cũng có thể chạy kiểm thử với tên của phép kiểm thử cụ thể(`test_foo`): 
```

File: src/cargo/test.md [hunk 16]
```diff
@@ -1 +1,134 @@
...
+    Finished dev [unoptimized + debuginfo] target(s) in 0.35 secs
+     Running target/debug/deps/blah-d3b32b97275ec472
+
+running 2 tests
+test test_foo ... ok
+test test_foo_bar ... ok
+
+test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 2 filtered out
+```
+Cảnh báo: Cargo có thể chạy nhiều kiểm thử cùng lúc, vì thế hãy chắc chắn rằng không xảy ra tranh chấp tài nguyên.
```
Comment (HongThaiPham) [inline src/cargo/test.md]:
```suggestion
Lưu ý: Cargo có thể chạy nhiều kiểm thử cùng lúc, vì thế hãy chắc chắn rằng không xảy ra tranh chấp tài nguyên. 
```

**Review by HongThaiPham — CHANGES_REQUESTED** [review]

File: src/cargo/test.md (hunk 16 above)
Comment (frenkie-ng) [reply src/cargo/test.md]: vâng ạ, em cảm ơn a!

**Review by nghuyenthevinh2000 — APPROVED** [review]

**Pull request merged**